quickwit/quickwit-metastore/src/metastore/retrying_metastore/mod.rs

```diff
@@ -0,0 +1,265 @@
+// Copyright (C) 2022 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+mod error;
+mod retry;
+#[cfg(test)]
+mod test;
+
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use quickwit_common::uri::Uri;
+use quickwit_config::SourceConfig;
+use quickwit_proto::metastore_api::{DeleteQuery, DeleteTask};
+
+use self::retry::{retry, RetryParams};
+use crate::checkpoint::IndexCheckpointDelta;
+use crate::{IndexMetadata, ListSplitsQuery, Metastore, MetastoreResult, Split, SplitMetadata};
+
+/// Retry layer for [`MetastoreGrpcClient`].
+/// This is a band-aid solution for now. This will be removed after retry can be usable on
+/// tonic level.
+/// Tracking Issue: https://github.com/tower-rs/tower/issues/682
+pub struct RetryingMetastore {
+    inner: Arc<dyn Metastore>,
```

Comment:
`RetryingMetastore` is always going to be wrapped in an `Arc`, right? Can we get away with only `Box` here then?